File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/TimeSegment.java
package nerd.tuxmobil.fahrplan.congress.schedule;

import info.metadude.android.eventfahrplan.commons.temporal.DateFormatter;
import info.metadude.android.eventfahrplan.commons.temporal.Moment;

/**
 * Represents a segment in the time column of the main schedule view.
 */
class TimeSegment {

    // TODO Merge definition with usage in FahrplanFragment
    private static final int TIME_GRID_MINIMUM_SEGMENT_HEIGHT = 5;
    private static final int MINUTES_PER_HOUR = 60;

    private final int hour;
    private final int minute;
    private final int minutesOfTheDay;

    /**
     * The given {@code minutesOfTheDay} are normalized. The minutes value is rounded to fit
     * into the time grid determined by {@code TIME_GRID_MINIMUM_SEGMENT_HEIGHT}.
     */
    TimeSegment(int minutesOfTheDay) {
        hour = minutesOfTheDay / MINUTES_PER_HOUR;
        minute = minutesOfTheDay % MINUTES_PER_HOUR;
        int remainder = minutesOfTheDay % TIME_GRID_MINIMUM_SEGMENT_HEIGHT;
        this.minutesOfTheDay = minutesOfTheDay - remainder;
    }

    /**
     * Returns the normalized and formatted text representing the given minutes of the day.
     * This text is ready to be displayed in the time column.
     */
    String getFormattedText() {
        Moment moment = Moment.now().startOfDay().plusMinutes(minutesOfTheDay);
        return DateFormatter.newInstance().getFormattedTime24Hour(moment);
    }

    boolean isMatched(Moment moment, int offset) {
        return moment.getHour() == hour && moment.getMinute() >= minute && moment.getMinute() < minute + offset;
    }

}

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/TimeSegment.kt
package nerd.tuxmobil.fahrplan.congress.schedule

import info.metadude.android.eventfahrplan.commons.temporal.DateFormatter
import info.metadude.android.eventfahrplan.commons.temporal.Moment
import nerd.tuxmobil.fahrplan.congress.schedule.TimeSegment.Companion.TIME_GRID_MINIMUM_SEGMENT_HEIGHT
import org.threeten.bp.ZoneOffset

/**
 * Represents a time column segment in the main schedule screen.
 *
 * The given [moment] is rounded down to the nearest grid boundary determined by
 * [TIME_GRID_MINIMUM_SEGMENT_HEIGHT]. The full date and time are retained so matching does
 * not accidentally include the same clock time on another day.
 *
 * The [formatted value][getFormattedText] is rendered in the time column in the main schedule view.
 */
internal class TimeSegment private constructor(moment: Moment) {

    companion object {

        /**
         * Represents both the minimum segment height and 5 minutes.
         */
        const val TIME_GRID_MINIMUM_SEGMENT_HEIGHT = 5

        fun ofMoment(moment: Moment) = TimeSegment(moment)

    }

    /**
     * Moment rounded down to the nearest time grid segment.
     *
     * Examples with a [TIME_GRID_MINIMUM_SEGMENT_HEIGHT] of 5 minutes:
     * - 10:00 remains 10:00.
     * - 10:04 becomes 10:00.
     * - 10:07 becomes 10:05.
     */
    private val roundedMoment: Moment

    init {
        val remainder = moment.minuteOfDay % TIME_GRID_MINIMUM_SEGMENT_HEIGHT
        roundedMoment = moment.minusMinutes(remainder.toLong())
    }

    /**
     * Returns the normalized and formatted text representing the given minutes of the day.
     * This text is ready to be displayed in the time column.
     *
     * See [DateFormatter.getFormattedTime24Hour].
     */
    fun getFormattedText(sessionZoneOffset: ZoneOffset?, useDeviceTimeZone: Boolean): String =
        DateFormatter.newInstance(useDeviceTimeZone).getFormattedTime24Hour(roundedMoment, sessionZoneOffset)

    /**
     * Returns true if the given [otherMoment] is inside the half-open interval
     * `[roundedMoment, roundedMoment + minutesOffset)`. Otherwise, false.
     * Consequently, a [minutesOffset] of 0 spans an empty interval which never matches.
     * An [IllegalArgumentException] is thrown if [minutesOffset] is negative.
     */
    fun isMatched(otherMoment: Moment, minutesOffset: Int): Boolean {
        require(minutesOffset >= 0) { "Minutes offset is $minutesOffset but must be 0 or more." }
        if (otherMoment.isBefore(roundedMoment)) {
            return false
        }
        return otherMoment.isBefore(roundedMoment.plusMinutes(minutesOffset.toLong()))
    }

    override fun toString() = roundedMoment.toString()

}
